### PR TITLE
Enhancement in install script to handle out of space errors

### PR DESCRIPTION
**Pull request by DivyaJyothiKasam**

### CODE REVIEW

**Comment by DivyaJyothiKasam** [comment]
Logs for RHEL73 with mlnx:
[root@rhel73 LISISO]# df -h
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/rhel-root   41G   41G  109M 100% /
devtmpfs               322M     0  322M   0% /dev
tmpfs                  333M     0  333M   0% /dev/shm
tmpfs                  333M  4.6M  328M   2% /run
tmpfs                  333M     0  333M   0% /sys/fs/cgroup
/dev/sda1             1014M  986M   29M  98% /boot
/dev/mapper/rhel-home   20G  1.2G   19G   6% /home
tmpfs                   67M     0   67M   0% /run/user/1000
[root@rhel73 LISISO]# ./install.sh
To install RPM you need to have minimum 30MB in boot and 140MB in root partition

[root@rhel73 LISISO]# df -Ph
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/rhel-root   41G   41G  108M 100% /
devtmpfs               322M     0  322M   0% /dev
tmpfs                  333M     0  333M   0% /dev/shm
tmpfs                  333M  4.6M  328M   2% /run
tmpfs                  333M     0  333M   0% /sys/fs/cgroup
/dev/sda1             1014M  986M   29M  98% /boot
/dev/mapper/rhel-home   20G  1.2G   19G   6% /home
tmpfs                   67M     0   67M   0% /run/user/1000
[root@rhel73 LISISO]# ./install.sh
To install RPM you need to have minimum 30MB in boot and 140MB in root partition

[root@rhel73 LISISO]# df -Ph
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/rhel-root   41G   41G  108M 100% /
devtmpfs               322M     0  322M   0% /dev
tmpfs                  333M     0  333M   0% /dev/shm
tmpfs                  333M  4.6M  328M   2% /run
tmpfs                  333M     0  333M   0% /sys/fs/cgroup
/dev/sda1             1014M  983M   32M  97% /boot
/dev/mapper/rhel-home   20G  1.2G   19G   6% /home
tmpfs                   67M     0   67M   0% /run/user/1000
[root@rhel73 LISISO]# ./install.sh
To install RPM you need to have minimum 30MB in boot and 140MB in root partition


[root@rhel73 LISISO]# df -Ph
Filesystem                    Size  Used Avail Use% Mounted on
/dev/mapper/rhel_rhel73-root   40G   40G  145M 100% /
devtmpfs                      320M     0  320M   0% /dev
tmpfs                         332M     0  332M   0% /dev/shm
tmpfs                         332M  5.0M  327M   2% /run
tmpfs                         332M     0  332M   0% /sys/fs/cgroup
/dev/sda1                    1014M  975M   40M  97% /boot
/dev/mapper/rhel_rhel73-home   20G  1.2G   19G   7% /home
tmpfs                          67M     0   67M   0% /run/user/1000
tmpfs                          67M     0   67M   0% /run/user/0
[root@rhel73 LISISO]# ./install.sh
Removing Hyper-V daemons
Removing Hyper-V Tools
Invoking release specific install file in directory RPMS73
RPM's are available
Installing the Linux Integration Services for Microsoft Hyper-V...
Preparing...                          ########################################
Updating / installing...
kmod-microsoft-hyper-v-4.3.0-20190404 ########################################
microsoft-hyper-v-4.3.0-20190404      ########################################
Saving old initramfs
Installing new initramfs
Starting KVP Daemon....
Starting VSS Daemon....
Starting FCOPY Daemon....
microsoft-hyper-v-debuginfo-4.3.0-2019########################################
 Linux Integration Services for Hyper-V has been installed. Please reboot your system.

**Comment by DivyaJyothiKasam** [comment]
Logs for RHEL75 without mlx:

[root@rhel75 LISISO]# df -h
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/rhel-root   41G   41G  109M 100% /
devtmpfs               322M     0  322M   0% /dev
tmpfs                  333M     0  333M   0% /dev/shm
tmpfs                  333M  4.6M  328M   2% /run
tmpfs                  333M     0  333M   0% /sys/fs/cgroup
/dev/sda1             1014M  986M   29M  98% /boot
/dev/mapper/rhel-home   20G  1.2G   19G   6% /home
tmpfs                   67M     0   67M   0% /run/user/1000
[root@rhel75 LISISO]# ./install.sh
To install RPM you need to have minimum 30MB in boot and 40MB in root partition

[root@rhel75 LISISO]# df -Ph
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/rhel-root   41G   41G  108M 100% /
devtmpfs               322M     0  322M   0% /dev
tmpfs                  333M     0  333M   0% /dev/shm
tmpfs                  333M  4.6M  328M   2% /run
tmpfs                  333M     0  333M   0% /sys/fs/cgroup
/dev/sda1             1014M  986M   29M  98% /boot
/dev/mapper/rhel-home   20G  1.2G   19G   6% /home
tmpfs                   67M     0   67M   0% /run/user/1000
[root@rhel75 LISISO]# ./install.sh
To install RPM you need to have minimum 30MB in boot and 40MB in root partition



[root@rhel75 LISISO]# df -Ph
Filesystem             Size  Used Avail Use% Mounted on
/dev/mapper/rhel-root   41G   41G  108M 100% /
devtmpfs               322M     0  322M   0% /dev
tmpfs                  333M     0  333M   0% /dev/shm
tmpfs                  333M  4.6M  328M   2% /run
tmpfs                  333M     0  333M   0% /sys/fs/cgroup
/dev/sda1             1014M  983M   32M  97% /boot
/dev/mapper/rhel-home   20G  1.2G   19G   6% /home
tmpfs                   67M     0   67M   0% /run/user/1000
[root@rhel75 LISISO]# ./install.sh
To install RPM you need to have minimum 30MB in boot and 40MB in root partition


[root@rhel75 LISISO]# df -Ph
Filesystem                    Size  Used Avail Use% Mounted on
/dev/mapper/rhel_rhel75-root   40G   40G  145M 100% /
devtmpfs                      320M     0  320M   0% /dev
tmpfs                         332M     0  332M   0% /dev/shm
tmpfs                         332M  5.0M  327M   2% /run
tmpfs                         332M     0  332M   0% /sys/fs/cgroup
/dev/sda1                    1014M  975M   40M  97% /boot
/dev/mapper/rhel_rhel75-home   20G  1.2G   19G   7% /home
tmpfs                          67M     0   67M   0% /run/user/1000
tmpfs                          67M     0   67M   0% /run/user/0
[root@rhel75 LISISO]# ./install.sh
Removing Hyper-V daemons
Removing Hyper-V Tools
Invoking release specific install file in directory RPMS75
RPM's are available
Installing the Linux Integration Services for Microsoft Hyper-V...
Preparing...                          ########################################
Updating / installing...
kmod-microsoft-hyper-v-4.3.0-20190404 ########################################
microsoft-hyper-v-4.3.0-20190404      ########################################
Saving old initramfs
Installing new initramfs
Starting KVP Daemon....
Starting VSS Daemon....
Starting FCOPY Daemon....
microsoft-hyper-v-debuginfo-4.3.0-2019########################################
 Linux Integration Services for Hyper-V has been installed. Please reboot your system.

**Comment by johnsongeorge-w** [comment]
@DivyaJyothiKasam : Define a variable to hold the MINIMUM space for boot partition and partition holding lib_modules and use it for comparison.

**Comment by DivyaJyothiKasam** [comment]
> @DivyaJyothiKasam : Define a variable to hold the MINIMUM space for boot partition and partition holding lib_modules and use it for comparison.

Added a variable

**Comment by DivyaJyothiKasam** [comment]
Linux Script Logs:
########CLEAN VM#################
[root@c73 LISISO]# ./install.sh
rpm_install_size 65.7301
existing_initramfs_size 20
min_initramfs 20
This is cleanup install
same_partition_minimum_req_disk 235.7301
diff_partition_minimum_req_disk_boot 40
diff_partition_minimum_req_disk_module 215.7301
in 73 same_partition_minimum_req_disk 285.7301
diff_partition_minimum_req_disk_module 265.7301
boot and modules do not share the same partition.
/dev/sda1 Required size: 40 MB. Available: 436
/dev/sda2 Required size: 265.7301 MB. Available: 27828
Removing Hyper-V daemons
Removing the hypervkvpd package
Removing the hyperv-daemons-license package
Removing Hyper-V Tools
Invoking release specific install file in directory RPMS73
Installing the Linux Integration Services for Microsoft Hyper-V...
Preparing...                          ########################################
Updating / installing...
kmod-microsoft-hyper-v-4.3.1-20190412 ########################################
microsoft-hyper-v-4.3.1-20190412      ########################################
Saving old initramfs
Installing new initramfs
Starting KVP Daemon....
Starting VSS Daemon....
Starting FCOPY Daemon....
microsoft-hyper-v-debuginfo-4.3.1-2019########################################
 Linux Integration Services for Hyper-V has been installed. Please reboot your system.

################UNINSTALL####################
[root@c73 LISISO]# ./uninstall.sh
Uninstalling the Linux Integration Services for Microsoft Hyper-V...
Removing Package..
Stopping KVP Daemon....
Stopping FCOPY Daemon....
Stopping VSS Daemon....
Create and install initramfs without Hyper-V drivers
Linux Integration Services for Hyper-V has been removed.
Uninstalled Linux Integration Services for Microsoft Hyper-V. Please reboot your system.


################UPGRADE##########################
rpm_install_size 65.7301
existing_initramfs_size 18
min_initramfs 0
This is re-install/upgrade
same_partition_minimum_req_disk 215.7301
diff_partition_minimum_req_disk_boot 20
diff_partition_minimum_req_disk_module 215.7301
in 73 same_partition_minimum_req_disk 265.7301
diff_partition_minimum_req_disk_module 265.7301
boot and modules do not share the same partition.
/dev/sda1 Required size: 20 MB. Available: 421
/dev/sda2 Required size: 265.7301 MB. Available: 27703
Removing Hyper-V daemons
Removing Hyper-V Tools
Invoking release specific install file in directory RPMS73
Installing the Linux Integration Services for Microsoft Hyper-V...
Preparing...                          ########################################
Updating / installing...
kmod-microsoft-hyper-v-4.3.1-20190412 ########################################
microsoft-hyper-v-4.3.1-20190412      ########################################
Saving old initramfs
Installing new initramfs
Starting KVP Daemon....
Starting VSS Daemon....
Starting FCOPY Daemon....
microsoft-hyper-v-debuginfo-4.3.1-2019########################################
 Linux Integration Services for Hyper-V has been installed. Please reboot your system.

File: LISISO/install.sh [hunk 1]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+temp_disk_usage_diff_partition_boot_clean_install=20
+
+#Reinstall Installation Temp Disk Usage
+temp_disk_usage_same_partition_reinstall=210
+temp_disk_usage_diff_partition_modules_reinstall=170
+temp_disk_usage_diff_partition_boot_reinstall=18
+
+#Calculations
+rpm_install_size=`rpm -qpi ${distro_package_name}${distro_version}/*hyper-v*x86_64.rpm | grep "Size" | awk '{ total += $3; print }; END { print total /1024/1024}' | tail -1`
+echo "rpm_install_size $rpm_install_size"
```
Comment (prashantmhetar) [inline LISISO/install.sh]: These debug prints need to be redirected to log file

File: LISISO/install.sh [hunk 2]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
  fi
 fi
 
+#Declarations
+
+mlx_driver_size=50
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: Is this required ? when we are quering the size of RPM which includes the mlx driver too.

File: LISISO/install.sh [hunk 3]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+
+hyperv_packages=$(rpm -qa | grep "hyper-v" | wc -l)
+if [ $hyperv_packages -le 0 ]; then
+        echo "This is cleanup install"
+        #Same Partition Calculation
+        same_partition_minimum_req_disk=$(bc <<< $rpm_install_size+$min_initramfs+$temp_disk_usage_same_partition_clean_install)
+        #Different Partition Calculation
+        diff_partition_minimum_req_disk_boot=$(expr $min_initramfs + $temp_disk_usage_diff_partition_boot_clean_install)
+        diff_partition_minimum_req_disk_module=$(bc <<< $rpm_install_size+$temp_disk_usage_diff_partition_modules_clean_install)
+        if [ "$distro_version" == 73 ] || [ "$distro_version" == 74 ]; then
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: Is this required ? when we are quering the size of RPM which includes the mlx driver too.

File: LISISO/install.sh [hunk 4]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+                diff_partition_minimum_req_disk_module=$(bc <<< $diff_partition_minimum_req_disk_module+$mlx_driver_size)
+        fi
+else
+        echo "This is re-install/upgrade"
+        #Same Partition Calculation
+        same_partition_minimum_req_disk=$(bc <<< $rpm_install_size+$min_initramfs+$temp_disk_usage_same_partition_reinstall)
+        #Different Partition Calculation
+        diff_partition_minimum_req_disk_boot=$(expr $min_initramfs + $temp_disk_usage_diff_partition_boot_reinstall)
+        diff_partition_minimum_req_disk_module=$(bc <<< $rpm_install_size+$temp_disk_usage_diff_partition_modules_reinstall)
+        if [ "$distro_version" == 73 ] || [ "$distro_version" == 74 ]; then
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: Is this required ? when we are quering the size of RPM which includes the mlx driver too.

File: LISISO/install.sh [hunk 5]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+        echo "$modules_partition_name Required size: $diff_partition_minimum_req_disk_module MB. Available: $current_modules_partition_space"
+        if [ ${current_boot_partition_space%.*} -lt ${diff_partition_minimum_req_disk_boot%.*} ]; then
+                echo "ERROR: Please make sure that you have sufficient disk space available in '$boot_partition_name'."
+                disk_space_errors=$(expr $disk_space_errors + 1)
+        fi
+        if [ ${current_modules_partition_space%.*}  -lt ${diff_partition_minimum_req_disk_module%.*} ]; then
+                echo "ERROR: Please make sure that you have sufficient disk space available in '$modules_partition_name'."
+                disk_space_errors=$(expr $disk_space_errors + 1)
+        fi
+fi
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: This should be implemented as a function and it should be just called before install

File: LISISO/install.sh [hunk 6]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+                disk_space_errors=$(expr $disk_space_errors + 1)
+        fi
+else
+        echo  "boot and modules do not share the same partition."
+        current_boot_partition_space=`df -hm | grep "/boot" | awk '{print $4}'`
+        current_modules_partition_space=`df -hm /lib/modules | awk '{print $4}' | tail -1`
+	echo "$boot_partition_name Required size: $diff_partition_minimum_req_disk_boot MB. Available: $current_boot_partition_space"
+        echo "$modules_partition_name Required size: $diff_partition_minimum_req_disk_module MB. Available: $current_modules_partition_space"
+        if [ ${current_boot_partition_space%.*} -lt ${diff_partition_minimum_req_disk_boot%.*} ]; then
+                echo "ERROR: Please make sure that you have sufficient disk space available in '$boot_partition_name'."
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: Looking for the following error message.

Not sufficient space to start LIS Installation.
                Partition containing /boot                       : Minimum Required Space (XX MB)      Current Space (YY MB)
                Partition containing lib_modules             : Minimum Required Space (ZZ MB)      Current Space (AA MB)

File: LISISO/install.sh [hunk 7]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+        echo "$boot_partition_name Required size: $same_partition_minimum_req_disk MB. Available: $current_boot_partition_size"
+        if [[ ( $current_boot_partition_size -lt $same_partition_minimum_req_disk ) ]]; then
+                echo "ERROR: Please make sure that you have sufficient disk space available in '$boot_partition_name'."
+                disk_space_errors=$(expr $disk_space_errors + 1)
+        fi
+else
+        echo  "boot and modules do not share the same partition."
+        current_boot_partition_space=`df -hm | grep "/boot" | awk '{print $4}'`
+        current_modules_partition_space=`df -hm /lib/modules | awk '{print $4}' | tail -1`
+	echo "$boot_partition_name Required size: $diff_partition_minimum_req_disk_boot MB. Available: $current_boot_partition_space"
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: This should only be printed when there is not sufficient space.

File: LISISO/install.sh [hunk 8]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+# Main script body
+#
+# Check if kernel version running is higher compared to installed kernel.
+# This is important because kmod are dependent on Kernel version
+
+disk_space_errors=0
+boot_partition_name=$(df -hm  /boot | awk '{print $1}' | tail -1)
+modules_partition_name=$(df -hm /lib/modules | awk '{print $1}' | tail -1)
+if [[ $boot_partition_name == $modules_partition_name ]]; then
+        echo  "boot and modules share the same partition."
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: These are internal and need not be printed for user.

File: LISISO/install.sh [hunk 9]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+temp_disk_usage_diff_partition_modules_clean_install=150
+temp_disk_usage_diff_partition_boot_clean_install=20
+
+#Reinstall Installation Temp Disk Usage
+temp_disk_usage_same_partition_reinstall=210
+temp_disk_usage_diff_partition_modules_reinstall=170
+temp_disk_usage_diff_partition_boot_reinstall=18
+
+#Calculations
+rpm_install_size=`rpm -qpi ${distro_package_name}${distro_version}/*hyper-v*x86_64.rpm | grep "Size" | awk '{ total += $3; print }; END { print total /1024/1024}' | tail -1`
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: As commented below. If this statement is inside a function and it is called just before install then distro_package_name and distro_version is not required.

File: LISISO/install.sh [hunk 10]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
  fi
 fi
 
+#Declarations
+
+mlx_driver_size=50
+
+#Clean Installation Temp Disk Usage
+temp_disk_usage_same_partition_clean_install=150
+temp_disk_usage_diff_partition_modules_clean_install=150
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: Why this is required ?

File: LISISO/install.sh [hunk 11]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+temp_disk_usage_diff_partition_boot_clean_install=20
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: Why this is required ?

File: LISISO/install.sh [hunk 12]
```diff
@@ -87,6 +81,94 @@ if [[ $distro_version != "5"* ]]; then
...
+
+#Reinstall Installation Temp Disk Usage
+temp_disk_usage_same_partition_reinstall=210
+temp_disk_usage_diff_partition_modules_reinstall=170
+temp_disk_usage_diff_partition_boot_reinstall=18
+
+#Calculations
+rpm_install_size=`rpm -qpi ${distro_package_name}${distro_version}/*hyper-v*x86_64.rpm | grep "Size" | awk '{ total += $3; print }; END { print total /1024/1024}' | tail -1`
+echo "rpm_install_size $rpm_install_size"
+existing_initramfs_size=`ls -lh /boot/"initramfs-$(uname -r).img" | awk '{print $5}' | tr -d "M"`
```
Comment (johnsongeorge-w) [inline LISISO/install.sh]: What will happen if the ramdisk file size is shown in KB or GB ? Dont use -h option for ls and get the size in bytes.

**Review by johnsongeorge-w — CHANGES_REQUESTED** [review]
Check the review comments and address accordingly.

**Pull request closed without merging**